raster/src/main/scala/geotrellis/raster/GridExtent.scala

```diff
@@ -70,8 +70,8 @@ class GridExtent[@specialized(Int, Long) N: Integral](
       throw GeoAttrsError(s"illegal cellheights: $cellheight and ${that.cellheight}")
 
     val newExtent = extent.combine(that.extent)
-    val newRows = ceil(newExtent.height / cellheight)
-    val newCols = ceil(newExtent.width / cellwidth)
+    val newRows = ceilWithTolerance(newExtent.height / cellheight)
```

Comment:
Should not be [here](https://github.com/locationtech/geotrellis/blob/4a021962d194b7693f8dff4b0cd9e10d49683c44/raster/src/main/scala/geotrellis/raster/GridExtent.scala#L45-L57) used ceilWithTolerance as well?

It looks like we're computing cols / rows via the ceil operation but validating it via the `round` function.

Can you also check that https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/MosaicRasterSource.scala#L160-L163 would work through the `GridExtent.combine`?

^ the last one I can check later; but if [this test](https://github.com/locationtech/geotrellis/blob/cf6e8596a0b21c4d3598525080b98dc087c8f898/gdal/src/test/scala/geotrellis/raster/gdal/GDALRasterSourceSpec.scala#L202) wont fail it is a good sign.

Reply:
> It looks like we're computing cols / rows via the ceil operation but validating it via the round function.

In short, yes, after a discussion with @echeipesh. And GridExtent currently operates in its constructor using `round` so in order to remain consistent, we should use that elsewhere when we construct new GridExtents, like in `withResolution` and `combine` to avoid mismatches throwing exceptions that are out of user control like in #3248. **I've updated this PR to use `round` everywhere and added an additional test to verify `withResolution`.** I've also created https://github.com/locationtech/geotrellis/issues/3261 in order to line up some future work to improve `withResolution`. I'm going to add a bit of documentation in here to describe this situation before merging as well.

> Can you also check...

I think we still need the logic there that you linked to for now because GridExtent.combine  will throw an exception unless the two GridExtents to be combined have the same CellSize, which isn't a guarantee here when constructing a mosaic.

Reply:
sounds good to me, thanks!